### PR TITLE
feat: Add support for error documents in the Agilent Gen5 adapter

### PR DESCRIPTION
**Pull request by joshua-benchling**
- Added support for error documents

### CODE REVIEW

File: src/allotropy/allotrope/schema_mappers/adm/plate_reader/rec/_2024/_06/plate_reader.py [hunk 1]
```diff
@@ -404,6 +437,17 @@ def _get_fluorescence_measurement_document(
...
+                error_document=[
+                    ErrorDocumentItem(
+                        error=error.error,
+                        error_feature=error.error_feature,
+                    )
+                    for error in measurement.error_document
+                ]
+            )
+            if measurement.error_document
+            else None,
```
Comment (slopez-b) [inline src/allotropy/allotrope/schema_mappers/adm/plate_reader/rec/_2024/_06/plate_reader.py:450]: Please create a function for instantiating the `ErrorAggregateDocument` to DRY.

File: tests/parsers/agilent_gen5/testdata/fluorescence/endpoint_single_plate_example.txt [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+																									Conc/Dil
+P	BLK	BLK	SPL8	SPL8	SPL16	SPL16	SPL24	SPL24	SPL32	SPL32	SPL40	SPL40	SPL48	SPL48	SPL56	SPL56	SPL64	SPL64	SPL72	SPL72	SPL80	SPL80	SPL88	SPL88	Well ID
+																									Conc/Dil
+
+
+
+Actual Temperature:	21.6
+
+Results
+	1	2	3	4	5	6	7	8	9	10	11	12	13	14	15	16	17	18	19	20	21	22	23	24
```
Comment (slopez-b) [inline tests/parsers/agilent_gen5/testdata/fluorescence/endpoint_single_plate_example.txt:73]: This file creates an ASM of 1.8Mb, since this is a new example and we already have plenty with large result tables, could you please reduce the sice of this matrix to 3x2? (3 columns, and keep `A` and `B` rows. Please do it with the Layout matrix too.

File: src/allotropy/parsers/agilent_gen5/constants.py [hunk 3]
```diff
@@ -39,6 +39,7 @@
 FILENAME_REGEX = r"^\d{6}_\d{6}_(?P<plate_identifier>.[^_]*)_.*\.txt$"
 
 NAN_EMISSION_EXCITATION = ["Full light", "Plug", "Hole"]
+NEGATIVE_ZERO = -0.0
```
Comment (slopez-b) [inline src/allotropy/parsers/agilent_gen5/constants.py:42]: I think if we are going to use `-0.0` to represent a bad reading of a measurement this constant should be in `src/allotropy/parsers/constants.py` for consistency.

cc @james-leinas

**Review by slopez-b — CHANGES_REQUESTED** [review]

File: src/allotropy/parsers/agilent_gen5/constants.py (hunk 3 above)
Comment (joshua-benchling) [reply src/allotropy/parsers/agilent_gen5/constants.py:42]: Yup, this makes sense. If we are ever going to apply to the other adapters, atleast for the ones for the new Plate Reader Schema.

File: tests/parsers/agilent_gen5/testdata/absorbance/absorbance_error_document_example.txt [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+Layout
+	1	2	3	4	5	6	7	8	9	10	11	12
+A	SPL1	SPL9	SPL17	SPL25	SPL33	SPL41	SPL49	SPL57	SPL65	SPL73	SPL81	SPL89	Well ID
+B	SPL2	SPL10	SPL18	SPL26	SPL34	SPL42	SPL50	SPL58	SPL66	SPL74	SPL82	SPL90	Well ID
+C	SPL3	SPL11	SPL19	SPL27	SPL35	SPL43	SPL51	SPL59	SPL67	SPL75	SPL83	SPL91	Well ID
+D	SPL4	SPL12	SPL20	SPL28	SPL36	SPL44	SPL52	SPL60	SPL68	SPL76	SPL84	SPL92	Well ID
+E	SPL5	SPL13	SPL21	SPL29	SPL37	SPL45	SPL53	SPL61	SPL69	SPL77	SPL85	SPL93	Well ID
+F	SPL6	SPL14	SPL22	SPL30	SPL38	SPL46	SPL54	SPL62	SPL70	SPL78	SPL86	SPL94	Well ID
+G	SPL7	SPL15	SPL23	SPL31	SPL39	SPL47	SPL55	SPL63	SPL71	SPL79	SPL87	BLK	Well ID
+H	SPL8	SPL16	SPL24	SPL32	SPL40	SPL48	SPL56	SPL64	SPL72	SPL80	SPL88	BLK	Well ID
```
Comment (slopez-b) [inline tests/parsers/agilent_gen5/testdata/absorbance/absorbance_error_document_example.txt:40]: These rows are not necessary since we only have results for row `A`

File: tests/parsers/agilent_gen5/testdata/fluorescence/endpoint_single_plate_example.json [hunk 5]
```diff
@@ -0,0 +1,648 @@
...
+                                "data source identifier": "AGILENT_GEN5_TEST_ID_0",
+                                "data source feature": "fluorescence"
+                            }
+                        ]
+                    }
+                },
+                {
+                    "calculated data name": "[Concentration]",
+                    "calculated result": {
+                        "value": -0.0,
```
Comment (slopez-b) [inline tests/parsers/agilent_gen5/testdata/fluorescence/endpoint_single_plate_example.json:457]: @james-leinas Should we still create a calculated data document for a value that was already reported in the error document? (`AGILENT_GEN5_TEST_ID_0` has an error document for the `[Concentration]` feature).

Comment (james-leinas) [reply tests/parsers/agilent_gen5/testdata/fluorescence/endpoint_single_plate_example.json:457]: @joshua-benchling, to @slopez-b point, we should disregard creating a calculated data document for values reported in the error document. Since the calculated data aggregate/calculated data document isn't required in the schema.

File: src/allotropy/parsers/agilent_gen5/constants.py (hunk 3 above)
Comment (james-leinas) [reply src/allotropy/parsers/agilent_gen5/constants.py]: @joshua-benchling, agree here! We'll implement the constant NEGATIVE_ZERO for each of the adapters for these scenarios of required field in the measurement.

File: src/allotropy/parsers/agilent_gen5/agilent_gen5_structure.py [hunk 6]
```diff
@@ -620,6 +634,7 @@ def create_results(
         )
         for well_position, well_calculated_data in calculated_data.items()
         for label, value in well_calculated_data
+        if value != NEGATIVE_ZERO
```
Comment (nathan-stender) [inline src/allotropy/parsers/agilent_gen5/agilent_gen5_structure.py:637]: I noticed that in the output files, we are dropping values that are `0`

This is because (unfortunately for us) in python `0 == -0`.

I think the way to solve this is to leave `well_value` as None and do `if value is not None` here, and then above where you set `value=value`, change it to `value=value if value is not None else NEGATIVE_ZERO` (note you will also have to check for `None` in `_create_measurement`)

**Review by nathan-stender — CHANGES_REQUESTED** [review]
Code mostly LGTM, but note we are dropping calculated data values when they are == 0, suggested a solution.

File: src/allotropy/parsers/agilent_gen5/agilent_gen5_structure.py (hunk 6 above)
Comment (joshua-benchling) [reply src/allotropy/parsers/agilent_gen5/agilent_gen5_structure.py:637]: It was really hard to follow this, but I think I got the idea. I managed to avoid dropping values that are 0 by comparing to None

File: src/allotropy/parsers/agilent_gen5/agilent_gen5_structure.py [hunk 7]
```diff
@@ -564,14 +566,25 @@ def create_results(
...
+        list[ErrorDocument]
+    )
     for row_name, row in data.iterrows():
         label = str(row.iloc[-1])
         for col_index, value in enumerate(row.iloc[:-1]):
             well_pos = f"{row_name}{col_index + 1}"
             well_value = try_non_nan_float_or_none(value)
-            # TODO: Report error documents for NaN values
-            if well_value is None:
+            if isinstance(value, float) and math.isnan(value):
```
[the remark this answers is not in the document]
Comment (joshua-benchling) [reply src/allotropy/parsers/agilent_gen5/agilent_gen5_structure.py:577]: you are right, I had to use `float_or_none`, so I can filter those nan values which are being return by an empty cell. In those cases we should not be reporting error documents.

File: src/allotropy/parsers/agilent_gen5/agilent_gen5_structure.py [hunk 8]
```diff
@@ -558,23 +560,44 @@ def create_results(
...
+
                 well_to_measurements[well_pos].append(
-                    MeasurementData(random_uuid_str(), well_value, label)
+                    MeasurementData(
+                        random_uuid_str(),
+                        NEGATIVE_ZERO if well_value is None else well_value,
+                        label,
+                    )
                 )
             else:
```
Comment (nathan-stender) [inline src/allotropy/parsers/agilent_gen5/agilent_gen5_structure.py:602]:
```suggestion
            else if value is not None:
```

One last suggestion, put this here, and remove it from below

File: src/allotropy/parsers/agilent_gen5/agilent_gen5_structure.py [hunk 9]
```diff
@@ -620,6 +644,7 @@ def create_results(
         )
         for well_position, well_calculated_data in calculated_data.items()
         for label, value in well_calculated_data
+        if value is not None
```
Comment (nathan-stender) [inline src/allotropy/parsers/agilent_gen5/agilent_gen5_structure.py:647]: Nit: put `None` check above, and remove this.

File: src/allotropy/parsers/agilent_gen5/agilent_gen5_structure.py [hunk 10]
```diff
@@ -558,23 +560,44 @@ def create_results(
     well_to_measurements: defaultdict[str, list[MeasurementData]] = defaultdict(
         list[MeasurementData]
     )
-    calculated_data: defaultdict[str, list[tuple[str, float]]] = defaultdict(
-        list[tuple[str, float]]
+    calculated_data: defaultdict[str, list[tuple[str, float | None]]] = defaultdict(
```
Comment (nathan-stender) [inline src/allotropy/parsers/agilent_gen5/agilent_gen5_structure.py:563]: By putting `None` check where I suggest below, you can change this back.

**Review by nathan-stender — CHANGES_REQUESTED** [review]
LGTM - one last nit

File: src/allotropy/parsers/agilent_gen5/agilent_gen5_structure.py (hunk 6 above)
Comment (nathan-stender) [reply src/allotropy/parsers/agilent_gen5/agilent_gen5_structure.py]: Sorry, but you got it!

**Review by nathan-stender — APPROVED** [review]
